File: src/gemini_voice/paths.py
import os
import platform
from pathlib import Path

from gemini_voice.config import BASE_DIR, DEFAULT_MODEL, load_config

MODELS_DIR = BASE_DIR / "models"
BIN_DIR = BASE_DIR / "bin"


def get_bin_path(bin_name: str) -> str | None:
    """Retorna o caminho do binário para o sistema operacional atual."""
    # Prioridade para variável de ambiente
    if bin_name == "piper":
        env_path = os.environ.get("VOICE_PIPER_PATH")
        if env_path and Path(env_path).exists():
            return env_path

    system = platform.system().lower()
    if system == "linux":
        path = BIN_DIR / "linux" / bin_name
    elif system == "windows":
        if not bin_name.endswith(".exe"):
            bin_name += ".exe"
        path = BIN_DIR / "windows" / bin_name
    else:
        return None

    if path.exists():
        return str(path)
    return None


def get_model_path() -> str | None:
    """Retorna o caminho do modelo de voz .onnx."""
    # Prioridade para variável de ambiente
    env_path = os.environ.get("VOICE_MODEL_PATH")
    if env_path and Path(env_path).exists():
        return env_path

    config = load_config()
    model_val = config.get("model", DEFAULT_MODEL)
    model_path = Path(model_val)

    if model_path.is_absolute() and model_path.exists():
        return str(model_path)

    # 1. Tentar diretamente como fornecido em MODELS_DIR
    relative_path = MODELS_DIR / model_val
    if relative_path.exists():
        return str(relative_path)

    # 2. Se falhar, tentar apenas o nome do arquivo se o caminho contiver diretórios (prefixo redundante)
    name_only = model_path.name
    name_path = MODELS_DIR / name_only
    if name_path.exists():
        return str(name_path)

    # 3. Fallback para o modelo padrão se o configurado falhar
    fallback_default = MODELS_DIR / DEFAULT_MODEL
    if fallback_default.exists():
        return str(fallback_default)

    if MODELS_DIR.exists():
        # Busca o primeiro arquivo .onnx disponível
        for f in sorted(MODELS_DIR.iterdir()):
            if f.suffix == ".onnx":
                return str(f)
    return None

File: src/paths.ts
import fs from "fs";
import path from "path";
import os from "os";
import { BIN_DIR, MODELS_DIR, DEFAULT_MODEL } from "./constants";
import { loadConfig } from "./config";

export function getBinPath(binName: string): string | null {
	const envPath = process.env.VOICE_PIPER_PATH;
	if (binName === "piper" && envPath && fs.existsSync(envPath)) {
		return envPath;
	}

	const system = os.platform();
	let fullBinName = binName;
	let subDir: string;

	if (system === "win32") {
		subDir = "windows";
		if (!fullBinName.endsWith(".exe")) fullBinName += ".exe";
	} else if (system === "linux") {
		subDir = "linux";
	} else {
		return null;
	}

	const binPath = path.join(BIN_DIR, subDir, fullBinName);
	return fs.existsSync(binPath) ? binPath : null;
}

export function getModelPath(): string | null {
	const envPath = process.env.VOICE_MODEL_PATH;
	if (envPath && fs.existsSync(envPath)) {
		return envPath;
	}

	const config = loadConfig();
	const modelVal = config.model || DEFAULT_MODEL;

	// 1. Try absolute path
	if (path.isAbsolute(modelVal) && fs.existsSync(modelVal)) {
		return modelVal;
	}

	// 2. Try extension's models directory
	const localPath = path.join(MODELS_DIR, modelVal);
	if (fs.existsSync(localPath)) {
		return localPath;
	}

	// 3. Try user's .voice/models directory
	const userModelDir = path.join(os.homedir(), ".voice", "models");
	const userPath = path.join(userModelDir, modelVal);
	if (fs.existsSync(userPath)) {
		return userPath;
	}

	// Try without extension if it fails (searching in both extension and user models dir)
	const nameOnly = path.basename(modelVal);
	const namePathLocal = path.join(MODELS_DIR, nameOnly);
	if (fs.existsSync(namePathLocal)) {
		return namePathLocal;
	}
	const namePathUser = path.join(userModelDir, nameOnly);
	if (fs.existsSync(namePathUser)) {
		return namePathUser;
	}

	// Fallback to default
	const fallbackDefault = path.join(MODELS_DIR, DEFAULT_MODEL);
	if (fs.existsSync(fallbackDefault)) {
		return fallbackDefault;
	}

	// Final search for any .onnx
	if (fs.existsSync(MODELS_DIR)) {
		const files = fs.readdirSync(MODELS_DIR);
		const onnxFile = files.find((f) => f.endsWith(".onnx"));
		if (onnxFile) {
			return path.join(MODELS_DIR, onnxFile);
		}
	}

	return null;
}
